sdk/cosmos/azure_data_cosmos/src/models/patch_operations.rs

```diff
@@ -0,0 +1,323 @@
+use azure_core::Model;
+use serde::{Deserialize, Serialize};
+
+// Cosmos' patch operations are _similar_ to JSON Patch (RFC 6902) in structure, but have different operations.
+
+/// Represents a patch document for Cosmos DB.
+///
+/// A patch document is made up of a collection of patch operations.
+/// Each operation describes how to modify a property of a document.
+/// See <https://learn.microsoft.com/en-us/azure/cosmos-db/partial-document-update> for more information on patch operations.
+///
+/// # Examples
+///
+/// To build up a patch document, use [`PatchDocument::default`] to create an empty document,
+/// then use the various `with_` methods to add operations.
+///
+/// ```rust
+/// # use azure_data_cosmos::models::{PatchDocument, PatchOperation};
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let patch = PatchDocument::default()
+///     .with_add("/color".to_string(), "silver")?
+///     .with_move("/from".to_string(), "/to".to_string())?;
+/// # assert_eq!(patch, PatchDocument {
+/// #     operations: vec![
+/// #         PatchOperation::Add {
+/// #             path: "/color".to_string(),
+/// #             value: serde_json::Value::String("silver".to_string()),
+/// #        },
+/// #        PatchOperation::Move {
+/// #            from: "/from".to_string(),
+/// #            to: "/to".to_string(),
+/// #        },
+/// #    ],
+/// # });
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Model, Default, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct PatchDocument {
+    pub operations: Vec<PatchOperation>,
+}
+
+impl PatchDocument {
+    /// Adds a new "add" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to add.
+    /// * `value` - The value to add.
+    pub fn with_add(
+        mut self,
+        path: String,
+        value: impl Serialize,
+    ) -> Result<Self, serde_json::Error> {
+        self.operations.push(PatchOperation::Add {
+            path,
+            value: serde_json::to_value(value)?,
+        });
+        Ok(self)
+    }
+
+    /// Adds a new "increment" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to increment.
+    /// * `value` - The amount to increment by. Integers can be specified directly in this parameter. Use [`serde_json::Number::from_f64`] to create a floating-point number.
+    pub fn with_increment(
+        mut self,
+        path: String,
+        value: serde_json::Number,
```

Comment:
This is another bit of an API conundrum. We want to accept `serde_json::Number` so we can support integers or floats. We could support `impl Into<serde_json::Number>` but there are two problems there:

1. We just added a guideline to avoid `impl` types. Though if the scenario is compelling, that doesn't bother me much
2. While all integer types can be converted `Into<Number>`, there are no implementations of `From<T> for Number` for any floating point types, because that conversion is fallible (JSON forbids NaN and Infinity). There are also no implementations of `TryFrom` for `Number` (though `serde_json` could add that)

So, for now I'm taking `Number` directly. We could also have our own trait, `impl` it for all the numeric primitives, and use it as an `impl` type here. I'm open to that, if we wanted to hide `serde_json::Number` from our API surface...

Reply:
I think I'd go with 1 in this case. Callers may not realize how to get a `serde_json::Number`, nor would I want to expose other crates' types in our public API (so, yes, maybe we do need our own trait like I did in https://github.com/heaths/ordinal-rs). That said, serde and friends are ubiquitous and may be safe.

Reply:
Newtonsoft.Json was once "ubiquitous and safe" too ;)

Reply:
I do think serde is even more entrenched though.

Reply:
A custom trait is easy enough and makes sense, IMO. Do we think this should be promoted up to `azure_core` or `typespec_client_core` or just live in the Cosmos SDK? I'll start with the latter for now, since we could move and re-export if necessary.

Reply:
> Newtonsoft.Json was once "ubiquitous and safe" too ;)
> I do think serde is even more entrenched though.

Certainly a valid concern, particularly given our own history with that particular situation, and the current struggles the .NET SDK is going through ;).

My sense from the Rust community is that serde has become a de-facto part of the "extended" stdlib and Rust is **much less** likely to bring serialization in to the real stdlib than .NET was. .NET included XML serialization from v1 (I believe it was there that early) whereas Rust has _no_ serialization in the stdlib.

The other element here is that Rust's Trait system is much more flexible than .NET's inheritance and interfaces. So even if a stdlib serialization system comes in, `serde_json` could easily adapt its types to support it **and** `serde`'s. That's a lot harder to do in .NET

It's definitely good to consider, but I think we're in a safer place than we're in a much better place depending on serde and friends than we were when we depended on Newtonsoft.Json.